models/topic.go

```diff
@@ -9,3 +11,45 @@ type Topic struct {
 	Keywords    []string `bson:"keywords,omitempty"     json:"keywords,omitempty"`
 	State       string   `bson:"state,omitempty"        json:"state,omitempty"`
 }
+
+// Validate checks that a topic struct complies with the state constraints, if provided. !!! may want to add more in future
+func (t *Topic) Validate() error {
+
+	if _, err := ParseState(t.State); err != nil {
+		return apierrors.ErrTopicInvalidState
+	}
+
+	// !!! add other checks, etc
+	return nil
+}
+
+// ValidateTransitionFrom checks that this topic state can be validly transitioned from the existing state
+func (t *Topic) ValidateTransitionFrom(existing *Topic) error {
+
+	// check that state transition is allowed, only if state is provided
+	if t.State != "" {
+		if !existing.StateTransitionAllowed(t.State) {
+			return apierrors.ErrTopicStateTransitionNotAllowed
+		}
+	}
+
+	// if the topic is already completed, it cannot be updated
+	if existing.State == StateCompleted.String() { //!!! ultimately this might not be needed
+		return apierrors.ErrTopicAlreadyCompleted
+	}
+
+	return nil
+}
+
+// StateTransitionAllowed checks if the topic can transition from its current state to the provided target state
+func (t *Topic) StateTransitionAllowed(target string) bool {
+	currentState, err := ParseState(t.State)
+	if err != nil {
+		currentState = StateCreated // default value, if state is not present or invalid value
+	}
+	targetState, err := ParseState(target)
+	if err != nil {
+		return false
```

Comment:
If this is an invalid state would we want to return the error?

Reply:
reminder comment added to code, to re-check once more of dp-topic-api implemented.